modules/programs/firefox/mkFirefoxModule.nix

```
@@ -775,47 +840,59 @@ in {
+    } ++ flip mapAttrsToList cfg.profiles (_: profile:
+      # Merge the regular profile settings with extension settings
+      mkMerge ([{
+        "${profilesPath}/${profile.path}/.keep".text = "";
+
+        "${profilesPath}/${profile.path}/chrome/userChrome.css" =
+          mkIf (profile.userChrome != "") { text = profile.userChrome; };
+
+        "${profilesPath}/${profile.path}/chrome/userContent.css" =
+          mkIf (profile.userContent != "") { text = profile.userContent; };
+
+        "${profilesPath}/${profile.path}/user.js" = mkIf (profile.preConfig
+          != "" || profile.settings != { } || profile.extraConfig != ""
+          || profile.bookmarks != [ ]) {
+            text =
+              mkUserJs profile.preConfig profile.settings profile.extraConfig
+              profile.bookmarks profile.extensions.settings;
+          };
 
-      "${profilesPath}/${profile.path}/containers.json" =
-        mkIf (profile.containers != { }) {
-          text = mkContainersJson profile.containers;
-          force = profile.containersForce;
-        };
+        "${profilesPath}/${profile.path}/containers.json" =
+          mkIf (profile.containers != { }) {
+            text = mkContainersJson profile.containers;
+            force = profile.containersForce;
+          };
 
-      "${profilesPath}/${profile.path}/search.json.mozlz4" =
-        mkIf (profile.search.enable) {
-          enable = profile.search.enable;
-          force = profile.search.force;
-          source = profile.search.file;
-        };
+        "${profilesPath}/${profile.path}/search.json.mozlz4" =
+          mkIf (profile.search.enable) {
+            enable = profile.search.enable;
+            force = profile.search.force;
+            source = profile.search.file;
+          };
 
-      "${profilesPath}/${profile.path}/extensions" =
-        mkIf (profile.extensions != [ ]) {
-          source = let
-            extensionsEnvPkg = pkgs.buildEnv {
-              name = "hm-firefox-extensions";
-              paths = profile.extensions;
-            };
-          in "${extensionsEnvPkg}/share/mozilla/${extensionPath}";
-          recursive = true;
-          force = true;
-        };
-    }));
+        "${profilesPath}/${profile.path}/extensions" =
+          mkIf (profile.extensions.packages != [ ]) {
+            source = let
+              extensionsEnvPkg = pkgs.buildEnv {
+                name = "hm-firefox-extensions";
+                paths = profile.extensions.packages;
+              };
+            in "${extensionsEnvPkg}/share/mozilla/${extensionPath}";
+            recursive = true;
+            force = true;
+          };
+      }] ++
+        # Add extension settings as separate attributes
+        optional (profile.extensions.settings != { }) (mkMerge (mapAttrsToList
+          (name: settingConfig: {
+            "${profilesPath}/${profile.path}/browser-extension-data/${name}/storage.js" =
+              {
+                force = true;
```

Do we really need to forcefully override this? Consider adding a `profiles.<profile>.extensions.settings.<extension>.force` option similar to the `profiles.<name>.search.force` one.

```suggestion
                force = settingsConfig.force;
```

---

I mean, we can, the issue arises that if firefox with extensions has been opened whatsoever, home-manager will fail to place the file

---

Reminds me of #3698.
> Its not enabled by default because it can wipe out an existing search configuration.

We should keep in mind that wiping out configurations is not the usual home manager behavior and should always be an active decision of the consuming user. However we can try to make that option as discoverable as possible.

---

fixed by https://github.com/nix-community/home-manager/pull/6389/commits/06b77de29a9f47aa9df90b646a19ea4f4ed34793